modules/create_timepoint/php/ajaxtimepoint.class.inc

```diff
@@ -0,0 +1,285 @@
+            case 'BadRequest':
+                return new \LORIS\Http\Response\BadRequest(
+                    'Failed to retrieve data for create timepoint.'
+                );
+            case 'Forbidden':
+                return new \LORIS\Http\Response\Forbidden(
+                    'Forbidden'
+                );
+            default:
+                break;
+            }
+        }
+
+        if (isset($values['command']) && $values['command'] == 'create') {
+            $errors = $this->validate($values);
+            // convert site entered to a \Site object
+            $site = null;
+            $user = \User::singleton();
+            $user_list_of_sites = $user->getCenterIDs();
+            $num_sites          = count($user_list_of_sites);
+            if ($num_sites == 1) {
+                $site = \Site::singleton($user_list_of_sites[0]);
+            } else if ($num_sites > 1) {
+                $site = \Site::singleton(intval($values['psc']));
+            }
+            if (!$errors && $site) {
+                \TimePoint::createNew(
+                    \Candidate::singleton($values['identifier']),
+                    intval($values['subproject']),
+                    $values['visit'],
+                    $site
+                );
+                $response['status'] = 'success';
+            } else {
+                $response['status'] = 'error';
+                $response['errors'] = $errors;
+            }
+        } else {
+            $response['status'] = 'error';
+        }
+        return new \LORIS\Http\Response\JsonResponse(
+            $response
+        );
+    }
+
+    /**
+     * Validate the post data,
+     * from the create timepoint form.
+     *
+     * @param array $values the form values.
+     *
+     * @return array
+     */
+    private function validate(array $values) : array
+    {
+        $user = \User::singleton();
+
+        $errors = array();
+
+        // validate site entered
+        $site = (int) $values['psc'];
+        $user_list_of_sites = $user->getData('CenterIDs');
+        $num_sites          = count($user_list_of_sites);
+        if ($num_sites > 1 && (empty($site) || !$user->hasCenter($site))) {
+            $errors['psc'] = 'Site must be selected from the available dropdown.';
+        }
+
+        $candid       = intval($values['identifier']);
+        $subprojectID = intval($values['subproject']);
+        $visitLabel   = $values['visit'] ?? '';
+
+        try {
+            \TimePoint::isValidVisitLabel($candid, $subprojectID, $visitLabel);
+        } catch (\LorisException $exception) {
+            $errors['visitLabel'] = $exception->getMessage();
+        }
+
+        return $errors;
+    }
+
+    /**
+     * Permission verification.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     * @param array                  $values  The form values.
+     *
+     * @return array
+     */
+    private function verifyPermissions(
```

Comment:
By convention, there should be `private function _hasAccess(\User $user): bool` that only check for permission.
Then, in the handlePOST, check for mandatory fields and return a BadRequest when one is missing.
Then, instantiate the candidate and compare its RegistrationCenterId to the user's sites. 

In it's current state, the `verifyPermissions` function is doing all this and have a side effect of modifying the values

Reply:
@xlecours this function is using the code before I modified the module for react. The $candidate is made from the $values['identifier] and that's used in the hasPermission('data_entry') check. Do you want me to move the validating the identifier sent from the fronted out of this function?

Reply:
@maltheism yes, I would like that.